gear-programs/erc20-relay/app/src/services/erc20_relay.rs

```diff
@@ -0,0 +1,274 @@
+        };
+
+        let (fungible_token, receipt, event) = self.prepare(&message)?;
+
+        let EthToVaraEvent {
+            proof_block: BlockInclusionProof { block, mut headers },
+            proof,
+            transaction_index,
+            ..
+        } = message;
+
+        // verify the proof of block inclusion
+        let checkpoints = self.state.borrow().checkpoints;
+        let slot = block.slot;
+        let checkpoint = Self::request_checkpoint(checkpoints, slot).await?;
+
+        headers.sort_unstable_by(|a, b| a.slot.cmp(&b.slot));
+        let Continue(block_root_parent) =
+            headers
+                .iter()
+                .rev()
+                .try_fold(checkpoint, |block_root_parent, header| {
+                    let block_root = header.tree_hash_root();
+                    match block_root == block_root_parent {
+                        true => Continue(header.parent_root),
+                        false => Break(()),
+                    }
+                })
+        else {
+            return Err(Error::InvalidBlockProof);
+        };
+
+        let block_root = block.tree_hash_root();
+        if block_root != block_root_parent {
+            return Err(Error::InvalidBlockProof);
+        }
+
+        // verify Merkle-PATRICIA proof
+        let receipts_root = H256::from(block.body.execution_payload.receipts_root.0 .0);
+        let mut memory_db = memory_db::new();
+        for proof_node in &proof {
+            memory_db.insert(hash_db::EMPTY_PREFIX, proof_node);
+        }
+
+        let trie = TrieDB::new(&memory_db, &receipts_root).map_err(|_| Error::TrieDbFailure)?;
+
+        let (key_db, value_db) =
+            eth_utils::rlp_encode_index_and_receipt(&transaction_index, &receipt);
+        match trie.get(&key_db) {
+            Ok(Some(found_value)) if found_value == value_db => (),
+            _ => return Err(Error::InvalidReceiptProof),
+        }
+
+        let amount = U256::from_little_endian(event.amount.as_le_slice());
+        let receiver = ActorId::from(event.to.0);
+        let call_payload =
+            vft_gateway::io::MintTokens::encode_call(fungible_token, receiver, amount);
+        let (reply_timeout, reply_deposit) = {
+            let state = self.state.borrow();
+
+            (state.reply_timeout, state.reply_deposit)
+        };
+        gstd::msg::send_bytes_for_reply(vft_gateway_id, call_payload, 0, reply_deposit)
+            .map_err(|_| Error::SendFailure)?
+            .up_to(Some(reply_timeout))
+            .map_err(|_| Error::ReplyTimeout)?
+            .handle_reply(move || handle_reply(slot, transaction_index))
+            .map_err(|_| Error::ReplyHook)?
+            .await
+            .map_err(|_| Error::ReplyFailure)?;
+
+        let _ = self.notify_on(Event::Relayed {
+            fungible_token,
+            to: ActorId::from(event.to.0),
+            amount,
+        });
+
+        Ok(())
+    }
+
+    fn prepare(
```

Comment:
It's not very clear from method name what are we preparing. Also struct for return type will be great